### ISSUE TITLE
REQUEST for uadmin headless version....

### ISSUE THREAD

**Issue by CharleGutierrez** [uadmin/uadmin#14]
Sir/Madam,
    Can you make another version of uadmin - headless, so we can run it on virtual container such as vagrant or docker for example, hope to see it work. :-)

**Comment by uadmin** [comment]
Hi @CharleGutierrez long time.

I didn't quite get your point about the headless implementation.

**Comment by CharleGutierrez** [comment]
can we separate the front end stuff or omit it only web services is expose to communicate with backend stuff... this is only my suggestion 👍 the consumer application will consume the web services expose.

**Comment by uadmin** [comment]
Sounds interesting. I think there is a big part that is missing from uAdmin which is a RESTful API to access the models data. I think that will really help as a first step for this.

**Comment by uadmin** [comment]
I have been looking at RESTful API as a standard for data access and I really hate one part of it. Every documentation of REST uses uncommon HTTP methods (PITCH, PUT and DELETE) for to send the CRUD command. This is an example and you will find many more:

https://www.oreilly.com/library/view/rest-in-practice/9781449383312/ch04.html#the_ordering_service_contract_overview

I really don't like the idea of using these HTTP methods as it will be a burden on most developers where GET and POST are usually fully supported in all languages and have short hand functions it almost all popular libraries (requests of python, JQuery of js, ...) so I'm wondering if we should just create a simpler standard that uses GET for reading and POST for writing. This is what I have on mind. Let's assume we have a model called `document`. As you register it you will be able to access it using these calls:

- The return format is always JSON in this format:
```json
{
  status: `ok`,
  items: {},
}
```

Where `value` has the return value of the call. Not all calls will have a return value. In case or error you will get a JSON result that looks like this:

```json
{
  status: `error`,
  err_msg: `error message`,
}
```

## Read:
 - Method: GET, POST
 - Call: http://api.exmaple.com/restapi/document/
 - Result: JSON array with all documents in the database
 - Auth: Cookie, URL parameter `?code=abcd` or POST parameter `code=abcd`

## Read One:
 - Method: GET, POST
 - Call: http://api.exmaple.com/restapi/document/1/
 - Result: JSON object representing document where `ID=1`
 - Auth: Cookie, URL parameter `?code=abcd` or POST parameter `code=abcd`

## Read Multiple:
 - Method: GET, POST
 - Call: http://api.exmaple.com/restapi/document/?name=contract
 - Result: JSON object representing document where `name="contract"`
 - Auth: Cookie, URL parameter `?code=abcd` or POST parameter `code=abcd`
 - Parameters: Any field of the model could be passed as a parameter to filter the results. Multiple filters could be used together where the system will apply `AND` operator between them. Evaluation operator could be added using `__` e.g. `?name__contains=contract`. This is the full this of operators:
  - `__contains` will search for string values that contract. Applies to: string
  - `__gt` greater than. Applies to: numbers, dates
  - `__gte` greater than or equal to. Applies to: numbers, dates
  - `__lt` less than. Applies to: numbers, dates
  - `__lte` less than or equal to. Applies to: numbers, dates
  - `__in` uses SQL `IN (a,c,b...)` operator to find a value matching any of these values. In the URL the values will be comma separated in this way `id__in=1,2,3`. Applies to: numbers
  - `__not` uses SQL `<>` which is used for not equal to.

## Add:
 - Method: GET, POST
 - Call: http://api.exmaple.com/restapi/document/add/
 - Result: No return value. Just `status`.
 - Auth: Cookie, URL parameter `?code=abcd` or POST parameter `code=abcd`
 - Parameters: The model's fields could be added as parameters depending of the HTTP method. If a field is required and there is no value passed, the API will return a `status:" 'error'` with an `err_msg` detailing which required field was missing.

## Edit:
 - Method: GET, POST
 - Call: http://api.exmaple.com/restapi/document/edit/1/
 - Result: No return value. Just `status`.
 - Auth: Cookie, URL parameter `?code=abcd` or POST parameter `code=abcd`
 - Parameters: Since the `ID` is already in the URL, you have to pass only the model's fields depending of the HTTP method.

## Delete:
 - Method: GET, POST
 - Call: http://api.exmaple.com/restapi/document/delete/1/
 - Result: No return value. Just `status`.
 - Auth: Cookie, URL parameter `?code=abcd` or POST parameter `code=abcd`

CC:
@Spotlight64 
jmsemira
@jgonito

Please give advice and feedback on the proposal.

**Comment by uadmin** [comment]
Missing standards for the API that needs further discussion and documentation:
 - OR operator
 - Limit, start, order
 - Selecting fields
 - Results meta data (limit, start, order, maybe some query execution time)
 - Delete multiple (using filters maybe?)
 - Edit multiple (using filters maybe?)
 - Add multiple
 - All API queries should run in one transaction
 - More operators:
  - `__startswith`
  - `__endswith`
  - `__re` regex
  - `__icontains` case insensitive
  - `__istartswith` case insensitive
  - `__iendswith` case insensitive
 - Model methods to control public access to the API
  - `func (m *Model) APIPublicRead() bool`
  - `func (m *Model) APIPublicAdd() bool`
  - `func (m *Model) APIPublicEdit() bool`
  - `func (m *Model) APIPublicDelete() bool`

Any thing missing? This is still a wishlist, so feel free to ask for any feature.

**Comment by uadmin** [comment]
Closed in version by  1afdcd60ef85ed6f55ba44f1c4c29bc94867eb8e
